misc/codepoint_width.c

```diff
@@ -0,0 +1,606 @@
+    0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x40, 0x42,
+    0x40, 0x40, 0x02, 0x02, 0x42, 0x02, 0x44, 0x42,
+    0x44, 0x42, 0x02, 0x40, 0x40, 0x40, 0x40, 0x02,
+    0x02, 0x42, 0x42, 0x42, 0x42, 0x02, 0x40, 0x40,
+    0x40, 0x42, 0x40, 0x40, 0x40, 0x02, 0x42, 0x44,
+    0x02, 0x02, 0x02, 0x02, 0x40, 0x40, 0x40, 0x40,
+    0x44, 0x44, 0x44, 0x44, 0x02, 0x02, 0x02, 0x02,
+    0x02, 0x02, 0x40, 0x42, 0x02, 0x02, 0x02, 0x02,
+    0x02, 0x02, 0x42, 0x02, 0x02, 0x42, 0x02, 0x02,
+    0x40, 0x02, 0x02, 0x02, 0x02, 0x02, 0x02, 0x40,
+    0x40, 0x02, 0x40, 0x02, 0x02, 0x40, 0x02, 0x02,
+    0x02, 0x02, 0x02, 0x02, 0x44, 0x02, 0x40, 0x40,
+    0x42, 0x42, 0x42, 0x42, 0x42, 0x40, 0x02, 0x02,
+    0x40, 0x42, 0x42, 0x02, 0x42, 0x02, 0x40, 0x40,
+    0x40, 0x02, 0x02, 0x42, 0x42, 0x40, 0x02, 0x02,
+    0x44, 0x42, 0x40, 0x40, 0x40, 0x40, 0x02, 0x02,
+    0x02, 0x40, 0x40, 0x40, 0x42, 0x42, 0x02, 0x42,
+    0x02, 0x40, 0x40, 0x40, 0x40, 0x40, 0x46, 0x40,
+    0x40, 0x40, 0x46, 0x46, 0x46, 0x40, 0x40, 0x40,
+    0x40, 0x40, 0x80, 0x80, 0x80, 0x80, 0x02, 0x40,
+    0x40, 0x40, 0x82, 0x82, 0x40, 0x40, 0x40, 0x40,
+    0x40, 0x40, 0x80, 0x80, 0x80, 0x80, 0x40, 0x80,
+    0x80, 0x40, 0x40, 0x80, 0x40, 0x40, 0x40, 0x40,
+    0x40, 0x80, 0x80, 0x80, 0x40, 0x40, 0x80, 0x40,
+    0x40, 0x01, 0x01, 0x01, 0x01, 0x40, 0x40, 0x40,
+    0x40, 0x42, 0x42, 0x42, 0x01, 0x01, 0x01, 0x01,
+    0x01, 0x02, 0x02, 0x02, 0x02, 0x02, 0x40, 0x40,
+    0x02, 0x02, 0x02, 0x02, 0x02, 0x4c, 0x4c, 0x4c,
+    0x4c, 0x8c, 0x4c, 0x4c, 0x4c, 0xc0, 0xc0, 0xc0,
+    0x40, 0x40, 0x4c, 0x4c, 0x4c, 0xc0, 0xc0, 0xc0,
+    0xc0, 0xc0, 0xc0, 0x40, 0x4c, 0xc0, 0xc0, 0x40,
+    0x40, 0x4c, 0x4c, 0x4c, 0x4c, 0xcc, 0xcc, 0xc0,
+    0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0xcc, 0xcc, 0xc0,
+    0xc0, 0xc0, 0xc0, 0xc0, 0xc0, 0x8c, 0xc0, 0x8c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0xc0, 0xc0,
+    0xc0, 0xc0, 0xc0, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c,
+    0x4c, 0x43, 0x43, 0x43, 0x43, 0x43, 0x43, 0x43,
+    0x43, 0x80, 0x8c, 0x8c, 0x4c, 0x4c, 0x4c, 0x4c,
+    0x4c, 0x80, 0x80, 0x8c, 0x80, 0x80, 0x80, 0x80,
+    0x80, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x80,
+    0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x8c,
+    0x8c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c, 0x4c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x4c, 0x8c,
+    0x8c, 0x8c, 0x4c, 0x4c, 0x4c, 0x4c, 0x8c, 0x4c,
+    0x4c, 0x4c, 0x8c, 0x4c, 0x4c, 0x4c, 0x8c, 0x8c,
+    0x8c, 0x82, 0x82, 0x82, 0x82, 0x82, 0x8c, 0x4c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x4c, 0x4c,
+    0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c, 0x40, 0x40,
+    0x4c, 0x4c, 0x4c, 0x8c, 0x8c, 0x8c, 0x8c, 0x4c,
+    0x40, 0x40, 0x40, 0x40, 0x4c, 0x4c, 0x4c, 0x4c,
+    0x4c, 0x4c, 0x40, 0x40, 0x40, 0x40, 0x8c, 0x8c,
+    0x8c, 0x8c, 0x40, 0x8c, 0x8c, 0x8c, 0x8c, 0x8c,
+    0x8c, 0x40, 0x8c, 0x41, 0x01, 0x41, 0x41, 0x41,
+    0x41, 0x41, 0x41, 0xc0, 0xc0, 0xc0, 0xc0, 0xc0,
+    0xc0, 0x40, 0x40,
+};
+
+int codepoint_width(const wchar_t cp)
+{
+    const uint16_t s0 = s_stage0[cp >> 11];
```

Comment:
On Windows `wchar_t` is 16 bits, which can't hold the codepoints of all the nice Emojies from the recent decade.

And on Linux where `wchar_t` is 32 bit, or everywhere if it's changed it to e.g. `uint32_t`, I'm quite sure that `s_stage0[cp >> 11]` can access out of bounds, because there are 21 bits remaining after shifting 11 of 32, and I doubt the array has 2M values, and if it doesn't then an arbitrary 21 bit index is not guaranteed to be in-bound.

In other words, this likely requires a test to ensure the input is within some range.

Reply:
> which can't hold the codepoints of all the nice Emojies from the recent decade.

They are "broken" on Windows anyway, because there wchar_t is used ubiquitously for wide character strings. But good point, no need to impose this limitation here.

> I doubt the array has 2M values, and if it doesn't then an arbitrary 21 bit index is not guaranteed to be in-bound.

As you probably know, 0x10FFFF is the maximum for valid code point. This is not external API, we already ensure only valid code points are extracted from utf-8 string.

> In other words, this likely requires a test to ensure the input is within some range.

It does not, it is on the caller.